File: Coding_practice/BOJ/1461/main.py
# 알고리즘 설계
# 1. 배열을 양수와 음수부로 나눈다
# 2. 배열의 양수와 음수부에서 최댓값을 계산하고, 큰 쪽을 마지막으로 가기
#   1) 양수/음수 둘 중 하나만 있는 경우에는 해당 방향을 마지막으로 가기.
#   2) 양수/음수 둘다 존재하는 경우에는 둘 중 최대값이 큰 쪽을 마지막으로 가기.
# 그리디 탐색 규칙 탐구
# 1. 기본 규칙
# 특정 방향으로 M개의 책을 가지고 가게되면, 가장 멀리 가야하는 책을 기준으로 M개의 책을 한번에 반환할 수 있기 때문에,
# max_dist * 2만큼의 거리만 있으면 M개의 책을 모두 제자리에 둘 수 있다.
# 예를 들어, [3, 10, 15, 20]이고 M == 3인 경우에는 20, 15, 10 책을 가지고 가면 40의 거리로 3개의 책을 제자리에 둘 수 있다.
# 2. 그리디 규칙
# 특정 방향을 마지막으로 가면, 해당 방향의 가장 큰 M개의 요소는 다시 돌아올 필요가 없기 때문에,
# *2가 아닌 *1한 값을 더해주면 된다.
# 양수부의 거리가 가장 먼 책과 음수부의 거리가 가장 먼 책 둘 중에서 더 멀리 있는 책을 왕복하지 않는게 유리하기 때문에,
# 양수부와 음수부의 max값을 비교하고 더 큰 방향에서 M개를 제외하고 step을 계산한다.
# 3. 예제
# [-39, -37, -29, -28, -6, 2, 11]
# -39, -37을 마지막에 둔다.
# 나머지 [-29, -28, -6, 2, 11]을 최소 거리로 두는 방법은,
# -29, -28을 가지고 가서 제자리에 놓고, -6을 가지고 가서 제자리에 놓고,
# 11, 2를 가지고 가서 제자리에 놓는 것임.
# 따라서, 39 + (29+6+11) * 2 = 131이 정답이다.
import sys
sys.stdin = open('input.txt', 'r')
sys.stdout = open('output.txt', 'w')


# flag=1인 경우에는 M개의 요소를 제외하고 step을 계산.
# flag=0인 경우에는 바로 step을 계산.
# 여기서 M개의 요소를 제외할 수 없는 경우가 걱정될 수 있는데, 그 경우에는 while의 i >= 0 조건에 바로 걸리기 때문에 문제 없음.
def get_steps(book_dist, flag):
    arr_len = len(book_dist)
    if flag:
        result = max(book_dist)
        i = arr_len - 1 - M
    else:
        result = 0
        i = arr_len - 1
    while i >= 0:
        result += book_dist[i] * 2
        i -= M
    return result


def solve():
    # 0. 기초 변수 할당
    ans = 0
    # 1. 배열을 양수/음수 나누기.
    arr_minus = [-x for x in arr if x < 0]
    arr_plus = [x for x in arr if x > 0]
    arr_minus.sort()
    arr_plus.sort()
    # 2. case 별로 값을 반환
    if not arr_minus:                   # 1) 음수가 없는 경우
        return get_steps(arr_plus, 1)
    elif not arr_plus:                  # 2) 양수가 없는 경우
        return get_steps(arr_minus, 1)
    else:                               # 3) 두 값을 비교해서 큰 쪽을 제외하기.
        minv = max(arr_minus)
        maxv = max(arr_plus)
        if maxv > minv:
            return get_steps(arr_plus, 1) + get_steps(arr_minus, 0)
        elif minv > maxv:
            return get_steps(arr_plus, 0) + get_steps(arr_minus, 1)


N, M = map(int, input().split())
arr = list(map(int, input().split()))

front = -1
rear = N-1
print(solve())

File: Coding_practice/BOJ/1461/Main.java
// 그리디 알고리즘 구상
// 1. 먼저 가장 멀리 있는 책을 한번만 가는게 좋아보인다.
// 가장 멀리 있는 책을 구하고, 그 책과 가장 가까운 M개의 책을 제자리에 놓을 때 한번만 가는게 좋다.
// 2. 나머지 책들을 반환하는 규칙은 어떻게 해야할까?
// 멀리 있는 책을 반환하면서 가장 가까이 있는 책들을 같이 가져가는게 좋다.
// [3, 4, 10, 11] 이렇게 책이 놓여야할 위치가 정해져 있고 M =2라면, 11을 반환할 때는 10, 11을 가지고 가서
// 한번에 반환하고 돌아오는게 좋다.
// 3. 양수와 음수를 같이 가져가는 것은 의미가 없다.
// 만약에 -2, 6책을 가지고 반납하러 간다면, 2 * 2 + 6 * 2 = 16만큼의 거리를 이동해야한다.
// 따라서 -2만 가지고 가서 반납한 뒤에 6만 가지고 가서 반납하는 것과 동일하다.
// 양수와 음수부를 나눠서 생각하는게 좋겠다.

// 알고리즘 구상
// 1. 자료 구조
// 먼저 배열을 양수부와 음수부로 나누고, 음수부는 절대값을 씌운다.
// 이후 두 배열을 정렬한다.
// 2. 탐색 시작
// 양수와 음수 둘 중에 하나가 없다면, 존재하는 쪽에서 최댓값을 찾고 그로부터 가까운 M개의 책을 마지막에 반납.
// 양수와 음수 둘다 있다면, 양수의 최댓값과 음수의 최댓값을 비교하여 더 큰 방향에서 M개의 책을 마지막에 반납.
// 반납하는 과정에서 거리를 계산하기.
import java.util.Arrays;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.FileInputStream;
import java.io.PrintStream;


class Main {
    static BufferedReader br;
    static String[] line;
    static int N, M;
    static String[] temp;
    static {
        try {
            // System.setIn(new FileInputStream("input.txt"));
            // System.setOut(new PrintStream("output.txt"));
            br = new BufferedReader(new InputStreamReader(System.in));
            line = br.readLine().split(" ");
            N = Integer.parseInt(line[0]);
            M = Integer.parseInt(line[1]);
            temp = br.readLine().split(" ");
        }
        catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static int getStep(int[] arr, int flag) {
        int lenArr = arr.length;
        int idx = lenArr - 1;
        int ans = 0;
        if (flag == 1) {
            ans += arr[idx];
            idx -= M;
        }
        while (idx >= 0) {
            ans += arr[idx]*2;
            idx -= M;
        }
        return ans;
    }
    public static void main(String[] args) {
        // 1. 입력값을 받고, 해당 배열을 양수부와 음수부로 나눈다.
        int[] pos = new int[N];
        int posTop = -1;
        int[] neg = new int[N];
        int negTop = -1;
        for (int i = 0; i < N; i ++) {
            int num = Integer.parseInt(temp[i]);
            if (num > 0) {
                posTop += 1;
                pos[posTop] = num;
            }
            else {
                negTop += 1;
                neg[negTop] = -num;
            }
        }
        Arrays.sort(pos);
        Arrays.sort(neg);
        // 2. 양수가 없는 경우, 음수가 없는 경우, 양수와 음수 둘 다 있는 경우.
        if (posTop == -1) {
            System.out.println(getStep(neg, 1));
        }
        else if (negTop == -1) {
            System.out.println(getStep(pos, 1));
        }
        else {
            int posMax = pos[pos.length-1];
            int negMax = neg[neg.length-1];
            if (posMax > negMax) {
                System.out.println(getStep(pos, 1)+getStep(neg, 0));
            }
            else {
                System.out.println(getStep(neg, 1)+getStep(pos, 0));
            }
        }
    }
}